logback/src/main/java/com/linecorp/armeria/common/logback/UnionMap.java

```diff
@@ -108,6 +109,26 @@ public void clear() {
             return entrySet;
         }
 
-        return entrySet = Collections.unmodifiableSet(Sets.union(first.entrySet(), second.entrySet()));
+        // Check for dupes first to reduce allocations on the vastly more common case where there aren't any.
+        boolean secondHasDupes = false;
+        for (Entry<K, V> entry : second.entrySet()) {
+            if (first.containsKey(entry.getKey())) {
+                secondHasDupes = true;
+                break;
+            }
+        }
+
+        final Set<Entry<K, V>> filteredSecond;
+        if (!secondHasDupes) {
+            filteredSecond = second.entrySet();
+        } else {
+            filteredSecond = new LinkedHashSet<>();
+            for (Entry<K, V> entry : second.entrySet()) {
+                if (!first.containsKey(entry.getKey())) {
+                    filteredSecond.add(entry);
+                }
+            }
+        }
+        return entrySet = Collections.unmodifiableSet(Sets.union(first.entrySet(), filteredSecond));
```

Comment:
IIUC, `Sets.union()` returns an immutable view, so there`s no need to wrap it again. It also handles the duplicate entries already, so maybe we can simplify most of this pull request?

Reply:
It handles duplicates, but the items of the set are Entry, not the key, so it's not deduping. Hence the bug in the first place :)

Reply:
Ah 🤣

Reply:
If that's the case, then maybe `ConcatenatedSet` is better, especially in terms of iteration performance?

Reply:
Ah yeah that's true, I forgot I removed the dedupe logic from it assuming the caller deduped to get some performance back. @kojilin Sorry for the bad pointer, it's probably good to avoid `Sets.union` here.